File: python/extensions/pybind_isce3/cuda/geometry/geo2rdr.cpp
#include "geo2rdr.h"

#include <string>

#include <isce3/core/Constants.h>
#include <isce3/core/Ellipsoid.h>
#include <isce3/core/LUT2d.h>
#include <isce3/core/Orbit.h>
#include <isce3/geometry/detail/Geo2Rdr.h>
#include <isce3/geometry/geometry.h>
#include <isce3/io/Raster.h>
#include <isce3/product/RadarGridParameters.h>

using isce3::cuda::geometry::Geo2rdr;

namespace py = pybind11;

void addbinding(py::class_<Geo2rdr> & pyGeo2Rdr)
{
    const isce3::geometry::detail::Geo2RdrParams defaults;
    pyGeo2Rdr
        .def(py::init([](const isce3::product::RadarGridParameters & radar_grid,
             const isce3::core::Orbit & orbit,
             const isce3::core::Ellipsoid & ellipsoid,
             const isce3::core::LUT2d<double> & doppler,
             const double threshold,
             const int numiter,
             const int lines_per_block)
            {
                auto geo2rdr_obj = Geo2rdr(radar_grid, orbit, ellipsoid, doppler);
                geo2rdr_obj.threshold(threshold);
                geo2rdr_obj.numiter(numiter);
                geo2rdr_obj.linesPerBlock(lines_per_block);
                return geo2rdr_obj;
            }),
            py::arg("radar_grid"),
            py::arg("orbit"),
            py::arg("ellipsoid"),
            py::arg("doppler") = isce3::core::LUT2d<double>(),
            py::arg("threshold") = defaults.threshold,
            py::arg("numiter") = defaults.maxiter,
            py::arg("lines_per_block") = 1000)
        .def("geo2rdr", py::overload_cast<isce3::io::Raster &, const std::string &,
                double, double>
                (&Geo2rdr::geo2rdr),
                py::arg("rdr2geo_raster"),
                py::arg("outdir"),
                py::arg("az_shift") = 0.0,
                py::arg("rg_shift") = 0.0)
        .def_property_readonly("orbit", &Geo2rdr::orbit)
        .def_property_readonly("ellipsoid", &Geo2rdr::ellipsoid)
        .def_property_readonly("doppler", &Geo2rdr::doppler)
        .def_property_readonly("radar_grid", &Geo2rdr::radarGridParameters)
        .def_property("threshold",
                py::overload_cast<>(&Geo2rdr::threshold, py::const_),
                py::overload_cast<double>(&Geo2rdr::threshold))
        .def_property("numiter",
                py::overload_cast<>(&Geo2rdr::numiter, py::const_),
                py::overload_cast<int>(&Geo2rdr::numiter))
        .def_property("lines_per_block",
                py::overload_cast<>(&Geo2rdr::linesPerBlock, py::const_),
                py::overload_cast<size_t>(&Geo2rdr::linesPerBlock))
        ;
}

File: tests/python/extensions/pybind/geometry/geo2rdr.py
#!/usr/bin/env python3

import os

from osgeo import gdal
import numpy as np
import numpy.testing as npt

import iscetest
import isce3.ext.isce3 as isce3
from nisar.products.readers import SLC

def test_point():
    h5_path = os.path.join(iscetest.data, "envisat.h5")

    radargrid = isce3.product.RadarGridParameters(h5_path)

    slc = SLC(hdf5file=h5_path)
    orbit = slc.getOrbit()
    doppler = slc.getDopplerCentroid()

    ellipsoid = isce3.core.Ellipsoid()

    llh = np.array([np.deg2rad(-115.72466801139711),
        np.deg2rad(34.65846532785868),
        1772.0])

    # run with native doppler
    aztime, slant_range = isce3.geometry.geo2rdr(llh, ellipsoid,
            orbit, doppler, radargrid.wavelength, radargrid.lookside,
            threshold=1.0e-10, maxiter=50, delta_range=10.0)

    azdate = orbit.reference_epoch + isce3.core.TimeDelta(aztime)
    assert azdate.isoformat() == "2003-02-26T17:55:33.993088889"
    npt.assert_almost_equal(slant_range, 830450.1859446081, decimal=6)

    # run again with zero doppler
    doppler = isce3.core.LUT2d()

    aztime, slant_range = isce3.geometry.geo2rdr(llh, ellipsoid,
            orbit, doppler, radargrid.wavelength, radargrid.lookside,
            threshold=1.0e-10, maxiter=50, delta_range=10.0)

    azdate = orbit.reference_epoch + isce3.core.TimeDelta(aztime)
    assert azdate.isoformat() == "2003-02-26T17:55:34.122893704"
    npt.assert_almost_equal(slant_range, 830449.6727720434, decimal=6)

    # try bracketing algorithm
    xyz = ellipsoid.lon_lat_to_xyz(llh)
    aztime, slant_range = isce3.geometry.geo2rdr_bracket(xyz, orbit,
            doppler, radargrid.wavelength, radargrid.lookside, tol_aztime=1e-8)

    azdate = orbit.reference_epoch + isce3.core.TimeDelta(aztime)
    assert azdate.isoformat() == "2003-02-26T17:55:34.122893704"
    npt.assert_almost_equal(slant_range, 830449.6727720434, decimal=6)

    # and again with a custom bracket
    aztime, slant_range = isce3.geometry.geo2rdr_bracket(xyz, orbit,
            doppler, radargrid.wavelength, radargrid.lookside, tol_aztime=1e-8,
            time_start=aztime - 1, time_end=aztime + 1)

    azdate = orbit.reference_epoch + isce3.core.TimeDelta(aztime)
    assert azdate.isoformat() == "2003-02-26T17:55:34.122893704"
    npt.assert_almost_equal(slant_range, 830449.6727720434, decimal=6)

def test_run():
    '''
    check if geo2rdr runs
    '''
    # prepare Geo2Rdr init params
    h5_path = os.path.join(iscetest.data, "envisat.h5")

    radargrid = isce3.product.RadarGridParameters(h5_path)

    slc = SLC(hdf5file=h5_path)
    orbit = slc.getOrbit()
    doppler = slc.getDopplerCentroid()

    ellipsoid = isce3.core.Ellipsoid()

    # init Geo2Rdr class
    tol_seconds = 1e-9
    tol_pixels = tol_seconds * radargrid.prf
    geo2rdr_obj = isce3.geometry.Geo2Rdr(radargrid, orbit,
            ellipsoid, doppler, threshold=tol_seconds, numiter=50)

    # load rdr2geo unit test output
    rdr2geo_raster = isce3.io.Raster("topo.vrt")

    # run
    geo2rdr_obj.geo2rdr(rdr2geo_raster, ".")

    # list of test outputs
    test_outputs = ["range.off", "azimuth.off"]

    # check each generated raster
    for test_output in test_outputs:
        # load dataset and get array
        test_ds = gdal.Open(test_output, gdal.GA_ReadOnly)
        test_arr = test_ds.GetRasterBand(1).ReadAsArray()

        # mask bad values
        test_arr = np.ma.masked_array(test_arr, mask=np.abs(test_arr) > 999.0)

        # compute max error (in pixels)
        test_err = np.max(np.abs(test_arr))

        # Allow a actual error to slightly exceed requested tolerance since
        # Newton step size isn't a perfect error estimate.
        assert (test_err < 2 * tol_pixels), f"{test_output} accumulated error fail"


if  __name__ == "__main__":
    test_run()
